design/design-doc-composition-functions-extra-resources.md

```diff
@@ -0,0 +1,1199 @@
+**Option 1**, allowing Function authors to expose whatever API they want to let
+Composition authors specify what resources they want, without needing to commit
+to any specific overly complex API. At least not immediately. We can always
+revisit this decision at a later stage if a perfect or "good enough" API arose
+from Functions and make it a native part of the Composition API.
+ 
+### Q1: How should Functions request additional resources?
+
+We considered the following options:
+- **Option 1**: Functions request additional resources through the
+  `RunFunctionResponse`.
+- **Option 2**: Functions request additional by calling back Crossplane through
+  a new RPC.
+
+We discarded **Option 2** for the following reasons:
+- it would require Crossplane to be a gRPC server too, which would be a
+  significant change to the current architecture, requiring Crossplane to be
+  reachable and addressable from Functions. This would, for example, make
+  `crossplane beta render` much more complex.
+- to be able to handle it securely we would need to implement some way for
+  Crossplane to check the "origin" of the request to validate that the requests
+  are authorized according to the policy defined in the Composition. To do so, we
+  would need to implement some sort of complex short-lived token mechanism.
+
+Therefore, **Option 1** is the only option we took into consideration.
+
+### Q2: Which steps in the pipeline should have access to the requested resources?
+
+We considered the following options:
+- **Option 1**: Only the Function that requested the resources should have
+  access to them.
+- **Option 2**: Only the next step in the pipeline should have access to the
+  requested resources.
+- **Option 3**: All the steps in the pipeline should have access to the
+  requested resources.
+- **Option 4**: Both the Function that requested the resources and the next
+  step in the pipeline should have access to the requested resources.
+
+We discarded **Option 3** because of the following reasons:
+- sending the requested resources to all the steps in the pipeline would
+  significantly increase network usage.
+- sending the requested resources to all the steps in the pipeline would
+  significantly increase the security risk as potentially sensitive information
+  could be leaked to steps that don't need it.
+- how would we handle multiple steps in the pipeline requesting resources? Would
+  Crossplane need to keep accumulating all the requested resources and continue
+  sending them to all the steps in the pipeline? Would be complex to define the
+  expected behavior.
+
+**Option 1**, **Option 2** and **Option 4** are all valid, we'll evaluate them
+in conjunction with the answer to **Q3**.
+
+### Q3: Should Functions be able to request additional resources incrementally?
+
+We considered the following options:
+- **Option 1**: Functions can request additional resources incrementally.
```

Comment:
It would be inefficient and Functions can always do that and we'll happily return all resources every time, I think it's an optimization we can implement from the beginning without too much additional effort.